optimizely/decision_service.py

```diff
@@ -340,10 +338,11 @@ def get_variation(
             self.logger.info(message)
             decide_reasons.append(message)
             # Store this new decision and return the variation for the user
-            if not ignore_user_profile and self.user_profile_service:
+            if user_profile_tracker is not None and not ignore_user_profile:
                 try:
-                    user_profile.save_variation_for_experiment(experiment.id, variation.id)
-                    self.user_profile_service.save(user_profile.__dict__)
+                    user_profile_tracker.update_user_profile(experiment, variation)
+                    if self.user_profile_service is not None:
+                        self.user_profile_service.save(user_profile_tracker.get_user_profile().__dict__)
```

Comment:
This will cause the user profile service to be saved for each flag key, making the batching save ineffective. This if block should be removed